vlm/src/main/scala/geotrellis/contrib/vlm/avro/GeoTrellisDataPath.scala

```diff
@@ -0,0 +1,71 @@
+package geotrellis.contrib.vlm.avro
+
+import geotrellis.contrib.vlm.DataPath
+
+
+/** Represents a path that points to a GeoTrellis layer saved in a catalog.
+ *
+ *  @param path Path to the layer. This can be either an Avro or COG layer.
+ *    The given path needs to be in a `URI` format that include the following query
+ *    parameters:
+ *      - '''layer''': The name of the layer.
+ *      - '''zoom''': The zoom level to be read.
+ *      - '''band_count''': The number of bands of each Tile in the layer.
+ *    Of the above three parameters, `layer` and `zoom` are required. In addition,
+ *    this path can be prefixed with, '''gt+''' to signify that the target path
+ *    is to be read in only by [[GeotrellisRasterSource]].
+ *  @example "s3://bucket/catalog?layer=layer_name&zoom=10"
+ *  @example "hdfs://data-folder/catalog?layer=name&zoom-12&band_count=5"
+ *  @example "gt+file:///tmp/catalog?layer=name&zoom=5"
+ *  @note The order of the query parameters does not matter.
+ */
+case class GeoTrellisDataPath(val path: String) extends DataPath {
+  private val servicePrefix: String = "gt+"
+
+  private val layerNameParam: String = "layer"
+  private val zoomLevelParam: String = "zoom"
+  private val bandCountParam: String = "band_count"
+
+  require(path.contains(s"$layerNameParam="), s"The layer query parameter must be in the given path: $path")
+
+  // TODO: Support having the zoom parameter be optional
+  require(path.contains(s"$zoomLevelParam="), s"The zoom query parameter must be in the given path: $path")
+
+  private val strippedPath: String =
+    if (path.startsWith(servicePrefix))
+      path.splitAt(servicePrefix.size)._2
+    else
+      path
+
+  val Array(catalogPath, queryParams) = strippedPath.split('?')
+
+  private val brokenUpParams: Array[(String, String)] =
```

Comment:
@pomadchin mentioned that there's a library called, [scala-uri](https://github.com/lemonlabsuk/scala-uri) that would handle the parsing for us; and it seems to do exactly what we're looking for. However, I'm not sure if it's worth bringing in another dependency to do this right now. If we start adding more complexity to the `URI`s, then I think that library should be something to consider.

Reply:
I would say that it would be pretty good to get rid of all that awful amount of regexps that make these `DataPath` classes unreadable. I created an issue for it: https://github.com/geotrellis/geotrellis-contrib/issues/192 

We can parse 1. schemes better 2. query params 3. authority 4. paths 5. encodes everything and supports lots of charsets. Basically we can get rid of the most `findFirstIn` conversions and lots of `requires` in code that break all the functional style of our library.